diff -r -U3 vendor_orig/aravis/src/arvgvinterface.c vendor/aravis/src/arvgvinterface.c
--- vendor_orig/aravis/src/arvgvinterface.c	2012-07-09 10:58:29.000000000 -0500
+++ vendor/aravis/src/arvgvinterface.c	2014-09-21 15:04:23.580825515 -0500
@@ -358,6 +358,76 @@
 	}
 }
 
+static GInetAddress * arv_gv_interface_camera_locate(ArvGvInterface *gv_interface, GInetAddress *device_address) {
+	ArvGvDiscoverSocketList *socket_list;
+	ArvGvcpPacket *packet;
+	GInetAddress *interface_address = NULL;
+	char buffer[ARV_GV_INTERFACE_SOCKET_BUFFER_SIZE];
+	GSList *iter;
+	size_t size;
+	int i, count;
+
+	socket_list = arv_gv_discover_socket_list_new ();
+
+	if (socket_list->n_sockets < 1) {
+		arv_gv_discover_socket_list_free (socket_list);
+		return interface_address;
+	}
+
+	/* Just read a random register from the camera socket */
+	packet = arv_gvcp_packet_new_read_register_cmd(ARV_GVBS_N_STREAM_CHANNELS_OFFSET, 0, &size);
+	GSocketAddress *device_socket_address = g_inet_socket_address_new(device_address, ARV_GVCP_PORT);
+
+	for (iter = socket_list->sockets; iter != NULL; iter = iter->next) {
+		ArvGvDiscoverSocket *socket = iter->data;
+		GError *error = NULL;
+		g_socket_send_to (socket->socket,
+				  device_socket_address,
+				  (const char *) packet, size,
+				  NULL, &error);
+		if (error != NULL) {
+			arv_warning_interface ("[ArvGVInterface::arv_gv_interface_camera_locate] Error: %s", error->message);
+			g_error_free (error);
+		}
+	}
+
+	arv_gvcp_packet_free (packet);
+
+	do {
+		/* Now parse the result */
+		if (g_poll (socket_list->poll_fds, socket_list->n_sockets, ARV_GV_INTERFACE_DISCOVERY_TIMEOUT_MS) == 0) {
+			arv_gv_discover_socket_list_free (socket_list);
+			return interface_address;
+		}
+
+		for (i = 0, iter = socket_list->sockets; iter != NULL; i++, iter = iter->next) {
+			ArvGvDiscoverSocket *socket = iter->data;
+
+			do {
+				g_socket_set_blocking (socket->socket, FALSE);
+				count = g_socket_receive (socket->socket, buffer, ARV_GV_INTERFACE_SOCKET_BUFFER_SIZE,
+							  NULL, NULL);
+				g_socket_set_blocking (socket->socket, TRUE);
+
+				if (count > 0) {
+					ArvGvcpPacket *packet = (ArvGvcpPacket *) buffer;
+					if (g_ntohs (packet->header.command) == ARV_GVCP_COMMAND_READ_REGISTER_CMD ||
+					    g_ntohs (packet->header.command) == ARV_GVCP_COMMAND_READ_REGISTER_ACK) {
+
+						interface_address = g_inet_socket_address_get_address(G_INET_SOCKET_ADDRESS (socket->interface_address));
+						g_object_ref(interface_address);
+						arv_gv_discover_socket_list_free (socket_list);
+						return interface_address;
+					}
+				}
+			} while (count > 0);
+		}
+	} while (1);
+	arv_gv_discover_socket_list_free (socket_list);
+	return interface_address;
+}
+
+
 static ArvDevice *
 arv_gv_interface_open_device (ArvInterface *interface, const char *device_id)
 {
@@ -381,6 +451,21 @@
 			device_infos = g_hash_table_lookup (gv_interface->priv->devices_by_mac, device_id);
 	}
 
+	if (device_infos == NULL) {
+	   /* This is probably an IP address */
+	   device_address = g_inet_address_new_from_string(device_id);
+	   if (device_address != NULL) {
+		   /* Try and find an interface that the camera will respond on */
+		   GInetAddress *interface_address = arv_gv_interface_camera_locate(gv_interface, device_address);
+		   if (interface_address != NULL) {
+			   device = arv_gv_device_new (interface_address, device_address);
+			   g_object_unref (interface_address);
+		   }
+		   g_object_unref (device_address);
+		   return device;
+	   }
+	}
+
 	if (device_infos == NULL)
 		return NULL;
 
diff -r -U3 vendor_orig/aravis/src/arvtool.c vendor/aravis/src/arvtool.c
--- vendor_orig/aravis/src/arvtool.c	2012-06-28 10:35:07.000000000 -0500
+++ vendor/aravis/src/arvtool.c	2014-09-21 15:04:45.683657085 -0500
@@ -145,7 +145,7 @@
 
 						if (ARV_IS_GC_STRING (feature))
 							printf ("%s = %s\n", tokens[0],
-								arv_gc_string_get_value (ARV_GC_STRING (feature), NULL));
+								arv_gc_feature_node_get_value_as_string (ARV_GC_FEATURE_NODE (feature), NULL));
 						else if (ARV_IS_GC_FLOAT (feature))
 							printf ("%s = %g (min:%g-max:%g)\n", tokens[0],
 								arv_gc_float_get_value (ARV_GC_FLOAT (feature), NULL),
